### PR TITLE
feat(eslint-plugin): add prefer-function-type rule

### PR DESCRIPTION
**Pull request by uniqueiniquity**
Add rule equivalent to TSLint's 'callable-types'.

### CODE REVIEW

File: packages/eslint-plugin/lib/rules/callable-types.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    /**
+     * Checks if there is no supertype or if the supertype is 'Function'
+     * @param {ESTreeNode} node The node being checked
+     * @returns {boolean} Returns true iff there is no supertype or if the supertype is 'Function'
+     */
+    function noSupertype(node) {
+      if (
+        typeof node.extends === 'undefined' &&
+        typeof node.implements === 'undefined'
```
Comment (bradzacher) [inline packages/eslint-plugin/lib/rules/callable-types.js]: better to check this as a boolean cast so it handles null case
```suggestion
        !node.extends && !node.implements
```

File: packages/eslint-plugin/lib/rules/callable-types.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      TSInterfaceDeclaration(node) {
+        if (noSupertype(node) && node.body.body.length === 1) {
+          checkMember(node.body.body[0], node);
+        }
+      },
+      /**
+       * @param {TSTypeLiteral} node The node being checked
+       * @returns {void}
+       */
+      TSTypeLiteral(node) {
```
Comment (bradzacher) [inline packages/eslint-plugin/lib/rules/callable-types.js]: Could change selector to remove the if:
```suggestion
      `TSTypeLiteral[members.length = 1]`(node) {
```

**Comment by bradzacher** [comment]
@armano2 / @JamesHenry - why do we "support" `implements` on interfaces?

Typescript errors out if you try it?
https://www.typescriptlang.org/play/#src=interface%20Bar%20%7B%0D%0A%0D%0A%7D%0D%0A%0D%0Ainterface%20Foo%20implements%20Bar%20%7B%0D%0A%20%20%0D%0A%7D

**Comment by armano2** [comment]
we errors on them to if user sets `errorOnTypeScriptSyntacticAndSemanticIssues`, right now we are outputting ast with "invalid" nodes to, this should not be handled in this rule

File: packages/eslint-plugin/lib/rules/callable-types.js (hunk 1 above)
Comment (armano2) [reply packages/eslint-plugin/lib/rules/callable-types.js]: you should check only `!node.extends`

**Comment by armano2** [comment]
you can find a little more info here: https://github.com/typescript-eslint/typescript-eslint/issues/185#issuecomment-460469085

**Comment by armano2** [comment]
if we are going to be able to get semantic errors in "cheap" way without creating program, i'm going to cleanup AST from all invalid nodes, right now prettier expects that we are going to output ast as is, and they can rebuild it without changing functionality only style

**Comment by j-f1** [comment]
Should this have a different name? Perhaps `prefer-function-type`?

**Comment by uniqueiniquity** [comment]
@j-f1 I'm open for naming suggestions. I was just matching TSLint's name, since I don't have a good sense of the ESLint rule naming conventions

**Comment by bradzacher** [comment]
(hint: we don't really have any solid naming conventions yet 😅 )

File: packages/eslint-plugin/lib/rules/callable-types.js (hunk 1 above)
Comment (uniqueiniquity) [reply packages/eslint-plugin/lib/rules/callable-types.js]: Changed to just check `extends`.

File: packages/eslint-plugin/lib/rules/callable-types.js (hunk 2 above)
Comment (uniqueiniquity) [reply packages/eslint-plugin/lib/rules/callable-types.js]: Great suggestion, didn't know that was doable. Fixed.

**Comment by bradzacher** [comment]
Code LGMT - change the name and we're good to go

**Comment by JamesHenry** [comment]
Apologies for the conflicts, we decided to remove the counts from the ROADMAP here: https://github.com/typescript-eslint/typescript-eslint/pull/225

They are too awkward to maintain manually, so this is the last time you will have to deal with syncing that up!

File: packages/eslint-plugin/lib/rules/prefer-function-type.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            : sourceCode
+                .getTokens(node)
+                .filter(
+                  token =>
+                    token.type === 'Keyword' && token.value === 'interface'
+                )[0].range[0];
+
+        context.report({
+          node: member,
+          messageId: 'callableTypeViolation',
```
Comment (JamesHenry) [inline packages/eslint-plugin/lib/rules/prefer-function-type.js]: Would be clearer to fix up the name here too

File: packages/eslint-plugin/lib/rules/prefer-function-type.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      description:
+        'Use function types instead of interfaces with call signatures',
+      category: 'TypeScript',
+      recommended: false,
+      extraDescription: [util.tslintRule('prefer-function-type')],
+      url: util.metaDocsUrl('prefer-function-type')
+    },
+    fixable: 'code',
+    messages: {
+      callableTypeViolation:
```
Comment (JamesHenry) [inline packages/eslint-plugin/lib/rules/prefer-function-type.js]: Same

File: packages/eslint-plugin/README.md [hunk 5]
```diff
@@ -143,6 +143,7 @@ Then you should add `airbnb` (or `airbnb-base`) to your `extends` section of `.e
 | [`@typescript-eslint/no-use-before-define`](./docs/rules/no-use-before-define.md)                         | Disallow the use of variables before they are defined                                                                                               | :heavy_check_mark: |          |
 | [`@typescript-eslint/no-useless-constructor`](./docs/rules/no-useless-constructor.md)                     | Disallow unnecessary constructors                                                                                                                   |                    |          |
 | [`@typescript-eslint/no-var-requires`](./docs/rules/no-var-requires.md)                                   | Disallows the use of require statements except in import statements (`no-var-requires` from TSLint)                                                 | :heavy_check_mark: |          |
+| [`@typescript-eslint/prefer-function-type`](./docs/rules/prefer-function-type.md)                         | Use function types instead of interfaces with call signatures (`prefer-function-type` from TSLint)                                                  |                    | :wrench: |
```
Comment (JamesHenry) [inline packages/eslint-plugin/README.md]: The `from TSLint` reference name should be reverted to its original TSLint name

Comment (uniqueiniquity) [reply packages/eslint-plugin/README.md]: Fixed.

File: packages/eslint-plugin/lib/rules/prefer-function-type.js (hunk 4 above)
Comment (uniqueiniquity) [reply packages/eslint-plugin/lib/rules/prefer-function-type.js]: Fixed.

File: packages/eslint-plugin/lib/rules/prefer-function-type.js (hunk 3 above)
Comment (uniqueiniquity) [reply packages/eslint-plugin/lib/rules/prefer-function-type.js]: Fixed.

**Pull request merged**